archetypal/umi_template.py

```diff
@@ -647,6 +647,26 @@ def to_dict(self):
 
         return data_dict
 
+    def alternate_unique_components(self, *args: str, exceptions: List[str] = None, keep_orphaned=False):
+        """ First, we divide the lib groups into equivalence classes"""
+        equivalence_classes = {}
+        for key, group in self:
+            eq_map = {}
+            for obj in group:
+                match_found = False
+                for exemplar in eq_map.values():
+                    if obj == exemplar:
+                        eq_map[obj] = exemplar
+                        match_found = True
+                        break
+                if not match_found:
+                    eq_map[obj] = obj
+            equivalence_classes[key] = eq_map
+        
+        for eq_class, eq_map in equivalence_classes.items():
+            for obj, equivalent_object in eq_map.items():
+                obj.replace_me_with(equivalent_object)
```

Comment:
Yeah, there are some issues that need to be sorted out.  I'm not 100% sure it's worth pursuing - the performance gains weren't as large as I was hoping when I did some benchmarking, but maybe the methods could be written slightly differently.  

Just did the `alternate` methods for now so that I could do the mentioned benchmarking.  

The main reason I'm not sure if it's worth implementing is because of the required additional `UmiBaseList` and `UmiBaseHelper` class which hook in to lists/wrapper classes respectively.  I think the approach is sound, but it adds a layer of complexity to the library since it is being done post-hoc that may not be worth it.  I'm a big fan of the ability to rapidly get a list of parent templates for any object, but I'm not sure if that functionality is enough to justify a fairly large refactor like this.

Thoughts?

Reply:
See #371 !